indico/web/client/js/react/components/DateRangePicker.jsx

```diff
@@ -0,0 +1,238 @@
+// This file is part of Indico.
+// Copyright (C) 2002 - 2024 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import moment from 'moment';
+import PropTypes from 'prop-types';
+import React from 'react';
+
+import DatePickerCalendar from 'indico/react/components/DatePickerCalendar';
+import {FinalField, validators as v} from 'indico/react/forms';
+import {Param, Translate} from 'indico/react/i18n';
+import {DateRange} from 'indico/utils/date';
+import {formatDate, ISO_FORMAT} from 'indico/utils/date_format';
+import {fromISOLocalDate} from 'indico/utils/date_parser';
+
+import 'indico/custom_elements/ind_date_picker';
+
+const INVALID = '__invalid__';
+
+export default function DateRangePicker({
+  value,
+  label,
+  format,
+  rangeStartLabel,
+  rangeEndLabel,
+  rangeStartMin,
+  rangeStartMax,
+  rangeEndMin,
+  rangeEndMax,
+  min,
+  max,
+  readOnly,
+  disabled,
+  onChange,
+}) {
+  format ??= moment.localeData().longDateFormat('L');
+
+  if (min) {
+    rangeStartMin = rangeStartMax = min;
+  }
+
+  if (max) {
+    rangeEndMin = rangeEndMax = max;
+  }
+
+  function handleChange(ev) {
+    const picker = ev.currentTarget;
+    const [start, end] = picker.dateRange;
+    const rangeStartInput = picker.querySelector('[data-range-start]');
+    const rangeEndInput = picker.querySelector('[data-range-end]');
+
+    const startInvalid = rangeStartInput.value && !start;
+    const endInvalid = rangeEndInput.value && !end;
+
+    onChange({
+      // ISO local dates
+      startDate: startInvalid ? INVALID : formatDate(ISO_FORMAT, start) || null,
+      endDate: endInvalid ? INVALID : formatDate(ISO_FORMAT, end) || null,
+    });
+  }
+
+  return (
+    <ind-date-range-picker
+      range-start={fromISOLocalDate(value.startDate)?.toDateString() || ''}
+      range-end={fromISOLocalDate(value.endDate)?.toDateString() || ''}
+      range-start-min={rangeStartMin}
+      range-start-max={rangeStartMax}
+      range-end-min={rangeEndMin}
+      range-end-max={rangeEndMax}
+      onChange={handleChange}
+    >
+      <fieldset>
+        <legend>{label}</legend>
+        <label>
+          <span>{rangeStartLabel}:</span>
+          <input
+            type="text"
+            data-range-start
+            placeholder={format}
+            readOnly={readOnly}
+            disabled={disabled}
+          />
+        </label>
+        <span className="arrow" />
+        <label>
+          <span>{rangeEndLabel}:</span>
+          <input
+            type="text"
+            data-range-end
+            placeholder={format}
+            readOnly={readOnly}
+            disabled={disabled}
+          />
+        </label>
+        <button type="button" data-calendar-trigger>
```

Comment:
There's one big change in behavior compared to the old date picker, and I wonder if that's going to be a problem.

When you already have a range selected in the old date range widget, you can open the calendar via either of the two fields, and the behavior slightly differs - in a way that's really convenient, especially when you have a long range:

- When opening it via the start date field, the first click will update the start date, but if you click on any date before the old end date, it will NOT clear the end date - so you can close the date picker and will have changed just the start date. When you select a date >= the end date, then it gets cleared.
- When opening it via the end date field, the behavior is similar: selecting anything after the start date will update the end date, and selecting something before the start date will update the start date (and then let you choose the end date)

I think this adds a huge amount of convenience, in particular when using the date range picker to edit something (e.g. an existing booking). You can easily have bookings that span multiple months (e.g. for a weekly meeting) and in that case it's quite important that you can simply change the end date without having to select the whole range again.

Not sure what's the best way to implement it though... maybe by having two calendar icons to open the picker from the start/end date?

Reply:
I think I can change the behavior so that:

1. When no dates are selected, the click selects the start date
2. When only one date is selected, the click selects the end date
3. When both dates are selected, the click modifies the one that's closer

That might cover your use case without modifying the existing UI.

What do you think?

Reply:
I think it covers most cases. Not sure about point 3, closer to what do you mean? To the current date? The distance between the one you clicked and the start/end dates? I think it's probably a good starting point and then when we can actually try it out we'll see if some more functionality is needed or not.

Reply:
By "close" I mean if you're clicking on a date that is closer to the start date, it will move the start date there, if it's closer to the end date, it will move the end date there.

Reply:
On a first try the behavior doesn't feel too bad (even though I guess it needs some getting-used-to). Any chance we could add a visual indicator during hovering to see what would be added/removed compared to the previous range?

Reply:
Yeah, I think I can do that. The reason I didn't is I figured it would be confusing if the mouse cursor is already somewhere in the calendar and it looks like nothing is selected.

Reply:
Done. Let me know if it works. Personally I find it a bit confusing as it behaves the same as if the range isn't selected, but I'm also too deep in this to make an objective judgement. 🤣

Reply:
I just tested it, and it makes it much clearer what happens.

But for both me ant @tomasr8 who also had a look, it also made it somewhat clear that it's a bit too much magic. In fact, even for me who knows exactly what the behavior is from our discussions here, it felt very clunky.

For a very simple range of let's say two weeks in the middle of a month it's not too bar - especially when extending it, it works. But you can't easily move e.g. from a two-week range to a one-week range the week after, since you're just changing the end date all the time.

So I'm afraid it'd be much better (especially for normal users) if we went back to what I suggested in my original comment. I think the only tricky part is how to nicely place a second trigger icon for the date picker in the start date field.
When doing that, would a11y be fine if we had some aria-live area showing you what date you  are currently selecting? Or can we assume that people using a screen reader would not use the calendar popup anyway but rather enter a date via the text field? In any case, some indicator - maybe even visible - what date you are currently selecting could be good for all users. But probably not necessary, at least I never encountered users not being able to select the range they waned in the old date range picker.

Reply:
> I think the only tricky part is how to nicely place a second trigger icon for the date picker in the start date field

So let me see if I understand this correctly.

Depending on which trigger you click, the selection only focuses on that end of the range. Is that the idea?

> Or can we assume that people using a screen reader would not use the calendar popup anyway

We can't. The calendar is fully instrumented to support screen readers and/or keyboard-only users.

Reply:
> Depending on which trigger you click, the selection only focuses on that end of the range. Is that the idea?

See what i wrote above ;)

> - When opening it via the start date field, the first click will update the start date, but if you click on any date before the old end date, it will NOT clear the end date - so you can close the date picker and will have changed just the start date. When you select a date >= the end date, then it gets cleared.
> - When opening it via the end date field, the behavior is similar: selecting anything after the start date will update the end date, and selecting something before the start date will update the start date (and then let you choose the end date)

So basically what you said, with some convenient functionality where operating on that end of the range would not make sense.

Reply:
> and selecting something before the start date will update the start date

So in this case it will behave like the latest version of my implementation?

Reply:
Yes, when you select something on the "left side" after having opened the date picker via the start date field, or on the "right side" in case of the end date field, it does exactly what your implementation does

Reply:
I'm currently enumerating all possible scenarios and matching them out to outcomes. I'll post the full list here once that's done. Meanwhile, do we want to allow selection of wrong ranges (where the order of the dates is swapped), and let the calendar automatically correct it?

Reply:
Ok. Let me try to recap.

In terms of interaction patterns, we have the following variables:

1. Selection state:

- no selection (N)
- left selected (L)
- right selected (R)
- both selected (B)

2. Calendar dialog trigger is used to target either the left or the right end of the range:

- open via left (L)
- open via right (R)

When one side is targeted, we call that end the 'target', and the other side the 'other'

3. User can make the following selections:

- Anywhere (A) - only applies when no selection is made
- Left of the target (target-left, TL)
- Right of the target (target-right, TR)
- Left of the other (other-left, OL)
- Right of the other (other-right OR)
- In-between the target and other (BT) - only when both ends are selected
- The target itself (T)
- The other itself (O)

There are optional range restrictions for both left and right targets, and we are going to assume that the following scenarios are allowed per the restrictions.

In the table the abbreviations in the outcome column mean:

- s. = select
- m. = move/update
- c. = clear

The abbreviations in the scenario and new state columns use two letters to describe the selection state and trigger used. For instance  LL means 'left only selected', 'left trigger used'. The part after the dash is the click target.

| left     | right    | trigger | click target  | scenario | outcome             | new state |
|----------|----------|---------|---------------|----------|---------------------|-----------|
| -        | -        | left    | anywhere      | NL-A     | s. left             | LR        |
| -        | -        | right   | anywhere      | NR-A     | s. right            | RL        |
|----------|----------|---------|---------------|----------|---------------------|-----------|
| selected | -        | left    | target        | LL-T     | c. left             | NL        |
| selected | -        | right   | other         | LR-O     | c. left             | NR        |
| selected | -        | left    | target-left   | LL-TL    | m. left             | LR        |
| selected | -        | left    | target-right  | LL-TR    | m. left             | LR        |
| selected | -        | right   | other-left    | LR-OL    | s. right + c. left  | RL        |
| selected | -        | right   | other-right   | LR-OR    | s. right + close    | -         |
|----------|----------|---------|---------------|----------|---------------------|-----------|
| -        | selected | right   | target        | RR-T     | c. right            | NR        |
| -        | selected | left    | other         | RL-O     | c. right            | NL        |
| -        | selected | right   | target-left   | RR-TL    | m. right            | RL        |
| -        | selected | right   | target-right  | RR-TR    | m. right            | RL        |
| -        | selected | left    | other-left    | RL-OL    | s. left + close     | -         |
| -        | selected | left    | other-right   | RL-OR    | s. left + c. right  | LR        |
|----------|----------|---------|---------------|----------|---------------------|-----------|
| selected | selected | left    | target        | BL-T     | c. left             | RL        |
| selected | selected | left    | other         | BL-O     | c. right            | LL        |
| selected | selected | left    | target-left   | BL-TL    | m. left             | BL        |
| selected | selected | left    | between       | BL-TW    | m. left             | BL        |
| selected | selected | left    | other-right   | BL-OR    | s. left + c. right* | LR        |
|----------|----------|---------|---------------|----------|---------------------|-----------|
| selected | selected | right   | target        | BR-T     | c. right            | LR        |
| selected | selected | right   | other         | BR-O     | c. left             | RR        |
| selected | selected | right   | target-right  | BR-TR    | m. right            | BR        |
| selected | selected | right   | between       | BR-TW    | m. right            | BR        |
| selected | selected | right   | other-left    | BR-OL    | s. left + c. right* | LR        |


\* this behavior is NOT mirrored for the other target, and the picker behaves the same regardless of which trigger is used

Please let me know if this sounds about right to you.

Reply:
> Meanwhile, do we want to allow selection of wrong ranges (where the order of the dates is swapped), and let the calendar automatically correct it?

Do you mean when you started selecting a start time, and then click on an earlier date? I kind of like the current behavior of replacing the start time in that case - I think it's more likely that you want to correct an incorrectly chosen start time in that case than wanting to treat the date as the end date instead.

Reply:
Ok, I've fixed the table for that particular case.

Reply:
I'm probably not the only one who has a hard time reading that table? So let's see if I do it correctly with the most simple case where nothing is preselected, and I click in the left field:

I guess it's the first row from the first group in the table (no selection + trigger left).
So I click on a date (A = anywhere) and it sets the left date (s,left).
Because the new state is LR, I now jump to the second group of the table, considering scenarios that start with `LR-`?

If I now click a date that's left to the previous selected one (start date), I get into the `LR-OL` scenario that updates the start date.
If I click a date that's right to the start date, I get into `LR-OR` which sets the end date and closes.

Does that sound right?

---

Now testing another case, I have two dates selected and click in the left field. So I consider `BL-` scenarios. I now click on a date left of the range, so I'm in `BL-TL` which means I modify the left/start date. However, the new state in the table is `BL`, but that sounds like it stays in the same mode (change start date). However, after having changed the start date the picker should switch to changing the end date, so I'd expect the new state to be `BR`...

Reply:
> Do I see it correctly that the left/right/trigger columns are just verbose versions of the first two characters of the "scenario"?

Yeah I think that is the case (also including `click target`), or at least I read the table the same way.

> However, after having changed the start date the picker should switch to changing the end date, so I'd expect the new state to be BR...

I also think it'd make more sense to switch to BR here

Reply:
> Does that sound right?

Correct.

> However, after having changed the start date the picker should switch to changing the end date, so I'd expect the new state to be BR...

Ok, I'll fix that. 

I've implemented this table as a set of functions already, so tweaking it in code should not be (that) hard. I've also realized that I can use the same code to "simulate" the next action when the user hovers the calendar for a an accurate representation of the final result. Just need to massage the code a bit to allow for this.

Reply:
Updated the PR with the changes as per the table + feedback.